subsys/bluetooth/controller/hci_int.c

```diff
@@ -0,0 +1,867 @@
+		BIT(22) | BIT(23) | BIT(24) | BIT(25) | BIT(28))
+
+#define ST_MAS2 (BIT(32 - 32) | BIT(33 - 32) | BIT(34 - 32) | BIT(35 - 32) | BIT(36 - 32) | \
+		BIT(37 - 32) | BIT(41 - 32))
+
+#if defined(CONFIG_BT_BROADCASTER)
+	states1 |= ST_ADV;
+#else
+	states1 &= ~ST_ADV;
+#endif
+#if defined(CONFIG_BT_OBSERVER)
+	states1 |= ST_SCA;
+#else
+	states1 &= ~ST_SCA;
+#endif
+#if defined(CONFIG_BT_PERIPHERAL)
+	states1 |= ST_SLA;
+	states2 |= ST_SLA2;
+#else
+	states1 &= ~ST_SLA;
+	states2 &= ~ST_SLA2;
+#endif
+#if defined(CONFIG_BT_CENTRAL)
+	states1 |= ST_MAS;
+	states2 |= ST_MAS;
+#else
+	states1 &= ~ST_MAS;
+	states2 &= ~ST_MAS2;
+#endif
+	/* All states and combinations supported except:
+	 * Initiating State + Passive Scanning
+	 * Initiating State + Active Scanning
+	 */
+	states1 &= ~(BIT(22) | BIT(23));
+	*buf = states1;
+	*(buf + 4) = states2;
+}
+
+#if defined(CONFIG_BT_CONN)
+static uint8_t link_control_cmd_put(uint8_t const * const cmd)
+{
+	uint16_t opcode = *(uint16_t *)cmd;
+	uint8_t const *cmd_params = &cmd[BT_HCI_CMD_HDR_SIZE];
+
+	switch (opcode)	{
+	case SDC_HCI_OPCODE_CMD_LC_DISCONNECT:
+		return sdc_hci_cmd_lc_disconnect((void *)cmd_params);
+	case SDC_HCI_OPCODE_CMD_LC_READ_REMOTE_VERSION_INFORMATION:
+		return sdc_hci_cmd_lc_read_remote_version_information((void *)cmd_params);
+	default:
+		return BT_HCI_ERR_UNKNOWN_CMD;
+	}
+}
+#endif
+
+static uint8_t controller_and_baseband_cmd_put(uint8_t const * const cmd,
+					       uint8_t * const raw_event_out,
+					       uint8_t *param_length_out)
+{
+	uint8_t const *cmd_params = &cmd[BT_HCI_CMD_HDR_SIZE];
+	uint16_t opcode = *(uint16_t *)cmd;
+#if defined(CONFIG_BT_CONN)
+	uint8_t * const event_out_params = &raw_event_out[CMD_COMPLETE_MIN_SIZE];
+#endif
+
+	switch (opcode)	{
+	case SDC_HCI_OPCODE_CMD_CB_SET_EVENT_MASK:
+		return sdc_hci_cmd_cb_set_event_mask((void *)cmd_params);
+	case SDC_HCI_OPCODE_CMD_CB_RESET:
+		return sdc_hci_cmd_cb_reset();
+
+#if defined(CONFIG_BT_CONN)
+	case SDC_HCI_OPCODE_CMD_CB_READ_TRANSMIT_POWER_LEVEL:
+		*param_length_out += sizeof(sdc_hci_cmd_cb_read_transmit_power_level_return_t);
+		return sdc_hci_cmd_cb_read_transmit_power_level((void *)cmd_params,
+								(void *)event_out_params);
+#endif
+#if defined(CONFIG_BT_HCI_ACL_FLOW_CONTROL)
+	case SDC_HCI_OPCODE_CMD_CB_SET_CONTROLLER_TO_HOST_FLOW_CONTROL:
+		return sdc_hci_cmd_cb_set_controller_to_host_flow_control((void *)cmd_params);
+	case SDC_HCI_OPCODE_CMD_CB_HOST_BUFFER_SIZE:
+		return sdc_hci_cmd_cb_host_buffer_size((void *)cmd_params);
+	case SDC_HCI_OPCODE_CMD_CB_HOST_NUMBER_OF_COMPLETED_PACKETS:
+		return sdc_hci_cmd_cb_host_number_of_completed_packets((void *)cmd_params);
+#endif
+
+#if defined(CONFIG_BT_CTLR_LE_ENC) && defined(BT_CTLR_LE_PING)
+	case SDC_HCI_OPCODE_CMD_CB_SET_EVENT_MASK_PAGE_2:
+		return sdc_hci_cmd_cb_set_event_mask_page_2((void *)cmd_params);
+	case SDC_HCI_OPCODE_CMD_CB_READ_AUTHENTICATED_PAYLOAD_TIMEOUT:
+		*param_length_out +=
```

Comment:
You could consider if net_buf_simple_init_with_data would have helped you here.
I find net_buf quite useful when it comes to this.

Reply:
- [ ] Use net_buf

Reply:
Consider it, not sure it would make it better, or smaller for that matter. :)